source/MRMesh/MRMeshBuilder.cpp

```diff
@@ -524,13 +528,46 @@ struct PathOverIncidentVert {
             {
                 --lastUnvisitedIndex;
                 std::iter_swap( it, vertexBegIt + lastUnvisitedIndex );
+                markVertexVisited(nextVertex);
                 return nextVertex;
             }
 
         }
         return VertId( -1 );
     }
 
+    // path = {abcDefgD} => closedPath = {DefgD}; path = {abc}
+    void extractClosedPath( std::vector<VertId>& path, std::vector<VertId>& closedPath )
+    {
+        auto lastVertex = path.back();
+        for ( size_t i = 0; i < path.size(); ++i )
+        {
+            if ( path[i] == lastVertex )
+            {
+                closedPath.reserve( path.size() - i );
+                closedPath.insert( closedPath.end(), std::make_move_iterator( path.begin() + i ),
+                                                     std::make_move_iterator( path.end() ) );
+
+                path.resize(i);
+                break;
+            }
+        }
+        for ( const auto& v : closedPath )
+            visitedVertices.reset( v );
```

Comment:
you can keep them set

Reply:
<img width="782" alt="Screenshot 2022-08-02 at 16 47 21" src="https://user-images.githubusercontent.com/46908529/182390835-f7421597-6dda-4a07-81ee-bede40df76fd.png">

Vertices are marked in visiting order. After duplicating 2-3-4-2, the algorithm will wrongly mark 1-2-5-3 for deduplication